pkg/cluster/advertise.go

```diff
@@ -7,31 +7,31 @@ import (
 	"github.com/pkg/errors"
 )
 
-// calculateAdvertiseAddress attempts to clone logic from deep within memberlist
+// calculateAdvertiseIP attempts to clone logic from deep within memberlist
 // (NetTransport.FinalAdvertiseAddr) in order to surface its conclusions to the
 // application, so we can provide more actionable error messages if the user has
 // inadvertantly misconfigured their cluster.
 //
 // https://github.com/hashicorp/memberlist/blob/022f081/net_transport.go#L126
-func calculateAdvertiseAddress(bindAddr, advertiseAddr string) (net.IP, error) {
-	if advertiseAddr != "" {
-		ip := net.ParseIP(advertiseAddr)
+func calculateAdvertiseIP(bindIP, advertiseIP string) (net.IP, error) {
+	if advertiseIP != "" {
+		ip := net.ParseIP(advertiseIP)
 		if ip == nil {
-			return nil, errors.Errorf("failed to parse advertise addr '%s'", advertiseAddr)
+			return nil, errors.Errorf("failed to parse advertise IP '%s'", advertiseIP)
 		}
 		if ip4 := ip.To4(); ip4 != nil {
 			ip = ip4
 		}
 		return ip, nil
 	}
 
-	if bindAddr == "0.0.0.0" {
+	if bindIP == "0.0.0.0" || bindIP == "::" || bindIP == "" {
 		privateIP, err := sockaddr.GetPrivateIP()
```

Comment:
On second thoughts, the private IP here should match the IP protocol implied by `bindIP`. It doesn't do this currently - the first private IP could be an IPv6 address.

Reply:
The problem with changes here in general is that this function, as the comment indicates, just replicates logic of memberlist to provide better error messages. Anything you change here won't actually happen when memberlist computes the address.

Reply:
Oh, I see - the IP address returned by `calculateAdvertiseAddress()` is only used for validation; the return value is not passed to the memberlist library.

I'll look at raising these changes upstream.